vispy/visuals/colorbar.py

```diff
@@ -436,75 +314,166 @@ def _update(self):
+            y = perp_direction[1]
+
+            perp_direction[0] = -y
+            perp_direction[1] = x
+
+        # use the document (pixel) coord system to set text anchors
+        anchors = []
+        if perp_direction[0] < 0:
+            anchors.append('right')
+        elif perp_direction[0] > 0:
+            anchors.append('left')
         else:
-            # raise an error since the orientation is now what was
-            # expected
-            raise ColorBarVisual._get_orientation_error(self._orientation)
+            anchors.append('center')
+        if perp_direction[1] < 0:
+            anchors.append('bottom')
+        elif perp_direction[1] > 0:
+            anchors.append('top')
+        else:
+            anchors.append('middle')
 
-    def _prepare_transforms(self, view):
-        print("preparing transorms")
-        CompoundVisual._prepare_transforms(view)
-        view._update()
+        return anchors
 
     @staticmethod
-    def _get_orientation_error(orientation):
-        return ValueError("orientation must"
-                          " be 'horizontal' or 'vertical', "
-                          "not '%s'" % (orientation, ))
+    def _calc_positions(center, halfdim, border_width,
+                        orientation, transforms):
+        """
+        Calculate the text centeritions given the ColorBar
+        parameters.
+
+        Note
+        ----
+        This is static because in principle, this
+        function does not need access to the state of the ColorBar
+        at all. It's a computation function that computes coordinate
+        transforms
+
+        Parameters
+        ----------
+        center: tuple (x, y)
+            Center of the ColorBar
+        halfdim: tuple (halfw, halfh)
+            Half of the dimensions measured from the center
+        border_width: float
+            Width of the border of the ColorBar
+        orientation: "top" | "bottom" | "left" | "right"
+            Position of the label with respect to the ColorBar
+        transforms: TransformSystem
+            the transforms of the ColorBar
+        """
+        (x, y) = center
+        (halfw, halfh) = halfdim
+
+        visual_to_doc = transforms.get_transform('visual', 'document')
+        doc_to_visual = transforms.get_transform('document', 'visual')
+
+        # doc_widths = visual_to_doc.map(np.array([halfw, halfh, 0, 0],
+        #                                         dtype=np.float32))
+
+        doc_x = visual_to_doc.map(np.array([halfw, 0, 0, 0], dtype=np.float32))
+        doc_y = visual_to_doc.map(np.array([0, halfh, 0, 0], dtype=np.float32))
+
+        if doc_x[0] < 0:
+            doc_x *= -1
+
+        if doc_y[1] < 0:
+            doc_y *= -1
+
+        # doc_halfw = np.abs(doc_widths[0])
+        # doc_halfh = np.abs(doc_widths[1])
 
-    @staticmethod
-    def _get_anchors(orientation):
         if orientation == "top":
-            return "center", "bottom"
+            doc_perp_vector = -doc_y
```

Comment:
Now, since we know the "canonical" directions in the document coordinates, figuring out the placement is dead simple. If you want to push the label "up", push it by `-doc_y`. Ditto with the other directions.